src/storage/src/client.rs

```diff
@@ -211,11 +224,21 @@ impl Storage {
                 "x-goog-api-client",
                 reqwest::header::HeaderValue::from_static(&self::info::X_GOOG_API_CLIENT_HEADER),
             );
-        let auth_headers = self
+        let cached_auth_headers = self
             .cred
             .headers(Extensions::new())
             .await
             .map_err(Error::authentication)?;
+        let auth_headers = match cached_auth_headers {
+            CacheableResource::New { data, .. } => Ok(data),
+            CacheableResource::NotModified => {
+                Err(Error::authentication(CredentialsError::from_str(
+                    false,
+                    "Auth headers not refreshed; client requires new headers to proceed with the request.",
+                )))
+            }
+        }?;
```

Comment:
Neat trick.  A bit easy to miss. Consider:

```suggestion
        };
   let auth_headers = auth_headers?;
```

Reply:
Separately, we need to refactor this code in the storage crate. Filed #2209